### PR TITLE
Add functional tests for sequential cut-off

### PR DESCRIPTION
**Pull request by micrakow**
Signed-off-by: Michal Rakowski <michal.rakowski@intel.com>

### CODE REVIEW

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+    """
+    title: Sequential cut-off tests during sequential and random IO 'always' policy with 4 cores
+    description: |
+        Testing if amount of data written to cache after sequential writes for different
+        sequential cut-off thresholds on each core, while running sequential IO on 3 out of 4
+        cores and random IO against the last core, is correct.
+    pass_criteria:
+        - Amount of written blocks to cache is less or equal than amount set
+          with sequential cut-off threshold for three first cores.
+        - Amount of written blocks to cache is equal sequential to io size run against last core.
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: This `sequential` word seems to be here by some kind accident.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+    """
+    with TestRun.step("Test prepare (start cache and add cores)"):
+        cache, cores = prepare(cores_count=4)
+        writes_before = []
+        io_sizes = []
+        thresholds = []
+        fio_additional_size = Size(10, Unit.Blocks4096)
+        for i in range(len(cores)):
+            thresholds.append(Size(random.randint(1, int((cache.cache_device.size
+                                                          * 0.5).get_value(Unit.KibiByte)))))
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: Please don't generate random parameters inside test case. This creates non-reproducible test cases. Use pytest parametrize instead (even if it's just oneshot per test case).

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        with TestRun.step(f"Setting core sequential cut off policy to {SeqCutOffPolicy.always}"):
+            core.set_seq_cutoff_policy(SeqCutOffPolicy.always)
+
+        with TestRun.step(f"Setting core sequential cut off threshold to {thresholds[i]}"):
+            core.set_seq_cutoff_threshold(thresholds[i])
+
+    with TestRun.group("Preparing FIO configuration"):
+        fio = (Fio().create_command()
+               .io_engine(IoEngine.libaio)
+               .direct())
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: This should be within step.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+
+        # Run sequential IO against first three cores
+        for i, core in TestRun.iteration(enumerate(cores[:-1])):
+            fio_job = fio.add_job(job_name=f"core_{core.core_id}")
+            fio_job.size(io_sizes[i])
+            fio_job.read_write(io_type)
+            fio_job.target(core.system_path)
+            writes_before.append(core.get_core_statistics(
+                stat_filter=[StatsFilter.blk])["writes to cache"])
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: This should be within step. BTW you don't need to use `TestRun.iteration()` everywhere. You can embed normal loop within step, especially when printing every iteration of the loop doesn't bring much information.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+
+        # Run random IO against the last core
+        fio_job = fio.add_job(job_name=f"core_{cores[-1].core_id}")
+        fio_job.size(io_sizes[-1])
+        fio_job.read_write(io_type_last)
+        fio_job.target(cores[-1].system_path)
+        writes_before.append(cores[-1].get_core_statistics(stat_filter=[StatsFilter.blk])[
+                             "writes to cache"])
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: This should be within step.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+
+    with TestRun.step("Running IO against all cores"):
+        fio.run()
+
+    with TestRun.step("Verifying writes to cache count after IO"):
+        for i, core in TestRun.iteration(enumerate(cores[:-1])):
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: You can't use iteration inside step. Make it step within iteration.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+#
+
+
+import pytest
+import random
+from enum import Enum, auto
+from api.cas import casadm
+from api.cas.core import SEQ_CUTOFF_THRESHOLD_MAX
+from api.cas.cache_config import SeqCutOffPolicy, CacheMode, CacheLineSize
+from api.cas.casadm import StatsFilter
```
Comment (Deixx) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: unused

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+    NEGATIVE = auto()
+    POSITIVE = auto()
+    EQUAL = auto()
+
+
+@pytest.mark.parametrize("thresholds_list", [[
+    random.randint(1, int(SEQ_CUTOFF_THRESHOLD_MAX.get_value(Unit.KibiByte))),
+    random.randint(1, int(SEQ_CUTOFF_THRESHOLD_MAX.get_value(Unit.KibiByte))),
+    random.randint(1, int(SEQ_CUTOFF_THRESHOLD_MAX.get_value(Unit.KibiByte))),
+    random.randint(1, int(SEQ_CUTOFF_THRESHOLD_MAX.get_value(Unit.KibiByte))),
```
[the remark this answers is not in the document]
Comment (micrakow) [reply test/functional/tests/seq_cutoff/test_seq_cutoff.py]: I did it for the CacheLineSize enum only, not for the Size class.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py (hunk 7 above)
Comment (micrakow) [reply test/functional/tests/seq_cutoff/test_seq_cutoff.py]: StatsFilter was indeed unused, removed.

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+    with TestRun.step("Verifying writes to cache count after IO"):
+        for i, core in enumerate(cores[:-1]):
+            verify_writes_count(core, writes_before[i], thresholds[i], io_sizes[i],
+                                VerifyType.POSITIVE)
+
+        verify_writes_count(cores[-1], writes_before[-1], thresholds[-1], io_sizes[-1],
+                            VerifyType.EQUAL)
+
+
+@pytest.mark.parametrize("threshold_param", [[
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: Do we need nested lists here while we want to pass just a single value?

File: test/functional/tests/seq_cutoff/test_seq_cutoff.py [hunk 10]
```diff
@@ -0,0 +1,263 @@
...
+              .read_write(io_dir)
+              .target(f"{cores[0].system_path}")
+              .direct()
+         ).run()
+
+    with TestRun.step("Verify writes to cache count"):
+        verify_writes_count(cores[0], writes_before, threshold, io_size, verify_type)
+
+
+@pytest.mark.parametrize("threshold_param", [[
```
Comment (robertbaldyga) [inline test/functional/tests/seq_cutoff/test_seq_cutoff.py]: Ditto.

**Pull request merged**